### DISCUSSION TITLE
Symfony 2 -> My website is not secured enough ?

### DISCUSSION THREAD

**Discussion by LudovicMaillet** [symfony/symfony#52470] [Q&A]
Hello,

I have a really big pb with a client's website... in Sympfony 2

It seems that the secret key configurerd to send mails using SMTP is used by hackers...
I think that maybe there is an access to the parameter.yml file...

The site : http://intranot.fr/

Could you please help me,
Maybe reading my security.yml and give a review?

Thanks for your help
Ludovic

**Comment by mbabker** [comment]
> Maybe reading my security.yml and give a review?

If anyone can get to this file on your server, you have a huge information disclosure issue as that would mean your entire site's configuration is publicly readable.  That being said, the fact that http://intranot.fr/ redirects me to http://intranot.fr/web/cms/accueil/ raises a huge potential red flag about what files from your application may be publicly accessible as that `web` folder should be the only thing the web server exposes.

**Reply by LudovicMaillet** [reply]
Thank you very much mbakker,

What steps should I follow to strengthen security?

Here is the security.yml content, do you see a pb ?

```
jms_security_extra:
    secure_all_services: false
    expressions: true

security:
    encoders:
        Sevanova\UserBundle\Entity\User: sha512

    role_hierarchy:
        ROLE_ADMIN:       ROLE_USER
        ROLE_SUPER_ADMIN: [ROLE_USER, ROLE_ADMIN, ROLE_ALLOWED_TO_SWITCH]

    providers:
        fos_userbundle:
            id: fos_user.user_manager

    firewalls:
        main:
            pattern:      ^/
            form_login:
                provider:       fos_userbundle
                login_path:     /cms/accueil/
                use_forward:    false
                check_path:     fos_user_security_check
                failure_path:   /cms/accueil/
                default_target_path: /cms/accueil/
            logout:
                path:   fos_user_security_logout
                target: /cms/accueil/
            anonymous:    true
            remember_me:
                key:      %secret%
                lifetime: 31536000
                path:     /
                domain:   ~
            switch_user: true
        dev:
            pattern:  ^/(_(profiler|wdt)|css|images|js)/
            security: false

    access_control:
        - { path: ^/login, role: IS_AUTHENTICATED_ANONYMOUSLY }
        - { path: ^/register, role: IS_AUTHENTICATED_ANONYMOUSLY }
        - { path: ^/resetting, role: IS_AUTHENTICATED_ANONYMOUSLY }
        - { path: ^/admin/, role: ROLE_ADMIN }
        - { path: ^/_profiler/, role: IS_AUTHENTICATED_ANONYMOUSLY }
        - { path: ^/.*, role: IS_AUTHENTICATED_ANONYMOUSLY }
```

**Reply by ghost** [reply]
The `security.yaml` contents has nothing to do here. It's "only" used for authentication and autorization. If someone has stolen your secret key then either it is (was?) outputed somewhere by mistake (or an exception occured in the code and leaked to the front with a sensitive parameter inside) or you don't have correctly configured web server for files access. You should only allow to access your public directory (which by default is `web` in SF 2.0 and `public` in SF 3 or 4).

**Reply by LudovicMaillet** [reply]
Do you mean using user permissions on directories, using server chmod  ?

**Reply by ghost** [reply]
I'm talking about web server configuration (Apache, Nginx etc.) where you declare what directories and files are allowed to be accessed from the web.

**Reply by LudovicMaillet** [reply]
Do you think it could be a wordaround to hide SMTP auth secert key by using sendmail with server config ? ...
Because even if I secure the web access, I'm not sure to fix the pb...
And I'm not a Symfony developer :/

**Reply by ghost** [reply]
I don't know.

**Reply by mbabker** [reply]
This isn't a Symfony issue.  This looks to be an Apache issue.  Only your project's `web/` folder should be accessible via the web server, not the entire project filesystem (which would include things like your root `composer.json`, or Composer's `vendor` folder).  So the biggest thing you need to do is change the Apache configuration so that ONLY the `web/` folder is exposed.

**Reply by sstok** [reply]
I think using Symfony 2 is already a problems, there are no longer any security patches, Symfony 2 is not compatible with PHP 8, which means it's properly an old PHP version which is properly also not secure.

This directory http://intranot.fr/vendor/ must not be publicly accessible, you can use a .htaccess file in directory you want to prevent access to.

```
# This will allow directory listing
Options -Indexes
```

But even better to use password protection (for any directory that must never be directly accessed), which is not the perfect solution but still better than nothing. https://stackoverflow.com/questions/5229656/password-protecting-a-directory-and-all-of-its-subfolders-using-htaccess

Tip: Use a password that cannot be validated (invalid hash), so that no one could possible guess the password.

**Reply by ghost** [reply]
@sstok 
> This directory http://intranot.fr/vendor/ must (not should) be publicly accessible, you can use a .htaccess file in directory you want to prevent access to.

**must not**

**Reply by LudovicMaillet** [reply]
I cannot change the config to only allow access to web, because this old website has been created like that... needing root of the website on its root folder
The agency which has designed it had a very bad idea...
I'm rebuilding it entirely, then it will be replaced by a Wordpress in a few monthes
Now I will secure all directories using .htaccess or a blank index.php

Thanks to you for your precious help
A beautiful thought for all of you from France
